packages/deployments/contracts/contracts/facets/StableSwapFacet.sol

```diff
@@ -0,0 +1,458 @@
+   * @param canonicalId the canonical token id
+   * @param tokenAmount the amount of the token you want to receive
+   * @param tokenIndex the index of the token you want to receive
+   * @param minAmount the minimum amount to withdraw, otherwise revert
+   * @param deadline latest timestamp to accept this transaction
+   * @return amount of chosen token user received
+   */
+  function removeStableLiquidityOneToken(
+    bytes32 canonicalId,
+    uint256 tokenAmount,
+    uint8 tokenIndex,
+    uint256 minAmount,
+    uint256 deadline
+  ) external nonReentrant deadlineCheck(deadline) returns (uint256) {
+    return s.swapStorages[canonicalId].removeLiquidityOneToken(tokenAmount, tokenIndex, minAmount);
+  }
+
+  /**
+   * @notice Remove liquidity from the pool, weighted differently than the
+   * pool's current balances. Withdraw fee that decays linearly
+   * over period of 4 weeks since last deposit will apply.
+   * @param canonicalId the canonical token id
+   * @param amounts how much of each token to withdraw
+   * @param maxBurnAmount the max LP token provider is willing to pay to
+   * remove liquidity. Useful as a front-running mitigation.
+   * @param deadline latest timestamp to accept this transaction
+   * @return amount of LP tokens burned
+   */
+  function removeStableLiquidityImbalance(
+    bytes32 canonicalId,
+    uint256[] calldata amounts,
+    uint256 maxBurnAmount,
+    uint256 deadline
+  ) external nonReentrant deadlineCheck(deadline) returns (uint256) {
+    return s.swapStorages[canonicalId].removeLiquidityImbalance(amounts, maxBurnAmount);
+  }
+
+  /*** ADMIN FUNCTIONS ***/
+  /**
+   * @notice Initializes this Swap contract with the given parameters.
+   * This will also clone a LPToken contract that represents users'
+   * LP positions. The owner of LPToken will be this contract - which means
+   * only this contract is allowed to mint/burn tokens.
+   *
+   * @param _canonicalId the canonical token id
+   * @param _pooledTokens an array of ERC20s this pool will accept
+   * @param decimals the decimals to use for each pooled token,
+   * eg 8 for WBTC. Cannot be larger than POOL_PRECISION_DECIMALS
+   * @param lpTokenName the long-form name of the token to be deployed
+   * @param lpTokenSymbol the short symbol for the token to be deployed
+   * @param _a the amplification coefficient * n * (n - 1). See the
+   * StableSwap paper for details
+   * @param _fee default swap fee to be initialized with
+   * @param _adminFee default adminFee to be initialized with
+   * @param lpTokenTargetAddress the address of an existing LPToken contract to use as a target
+   */
+  function initializeStableSwap(
+    bytes32 _canonicalId,
+    IERC20[] memory _pooledTokens,
+    uint8[] memory decimals,
+    string memory lpTokenName,
+    string memory lpTokenSymbol,
+    uint256 _a,
+    uint256 _fee,
+    uint256 _adminFee,
+    address lpTokenTargetAddress
+  ) external onlyOwner {
```

Comment:
needs ``initializer` modifier from openzeppelins `Initializable`, no?

otherwise, even though this is limited by `onlyOwner`, it can be reinitialized maliciously

Reply:
`initializer` modifier is for only `Initializable`. 

to prevent that, we have this.
`if (s.swapStorages[_canonicalId].pooledTokens.length != 0)
      revert StableSwapFacet__initializeSwap_alreadyInitialized();`

Reply:
oh yeah, just realized the structure is different from typical stableswap in that this contract manages many pools indexed by canonical id

sounds good